File: apply_migration_12.js
#!/usr/bin/env node
/**
 * Apply migration 12: Fix material capacity MV unique index
 * This fixes the 500 error when updating expected materials.
 */

const fs = require('fs');
const path = require('path');
require('dotenv').config();

const { createClient } = require('@supabase/supabase-js');

async function main() {
  // Get Supabase credentials
  const supabaseUrl = process.env.SUPABASE_URL;
  const serviceKey = process.env.SUPABASE_SERVICE_KEY;

  if (!supabaseUrl || !serviceKey) {
    console.error('Error: Missing Supabase credentials in .env file');
    console.error('Required: SUPABASE_URL and SUPABASE_SERVICE_KEY');
    process.exit(1);
  }

  // Create Supabase client with service role
  const supabase = createClient(supabaseUrl, serviceKey, {
    auth: {
      autoRefreshToken: false,
      persistSession: false
    }
  });

  // Read migration SQL
  const migrationFile = path.join(__dirname, 'supabase/migration/12_fix_material_capacity_unique_index.sql');
  const sql = fs.readFileSync(migrationFile, 'utf8');

  console.log('Executing migration 12...');
  console.log('Project:', supabaseUrl.replace('https://', '').split('.')[0]);

  try {
    // Split into individual SQL statements and execute them
    const statements = sql
      .split(';')
      .map(s => s.trim())
      .filter(s => s && !s.startsWith('--') && !s.match(/^\/\*/));

    for (let i = 0; i < statements.length; i++) {
      const stmt = statements[i];
      if (!stmt) continue;

      console.log(`\nExecuting statement ${i + 1}/${statements.length}...`);

      const { data, error } = await supabase.rpc('exec_sql', {
        sql_query: stmt + ';'
      });

      if (error) {
        // If exec_sql function doesn't exist, provide alternative instructions
        if (error.message && error.message.includes('could not find function')) {
          console.log('\nThe exec_sql function is not available in your database.');
          console.log('\nPlease execute the migration manually using one of these methods:\n');
          console.log('1. Supabase Dashboard SQL Editor:');
          console.log(`   - Go to: ${supabaseUrl.replace('https://', 'https://supabase.com/dashboard/project/')}/sql`);
          console.log('   - Copy and paste the contents of: supabase/migration/12_fix_material_capacity_unique_index.sql');
          console.log('   - Click "Run"\n');
          console.log('2. Use the SQL directly (copy from the file above)');
          process.exit(1);
        }
        throw error;
      }
    }

    console.log('\n✓ Migration executed successfully!');
    console.log('The 500 error should now be fixed.');
    console.log('\nYou can now update expected materials without errors.');

  } catch (error) {
    console.error('\n✗ Error executing migration:', error.message);
    console.log('\nPlease execute manually via Supabase Dashboard SQL Editor:');
    console.log(`https://supabase.com/dashboard/project/${supabaseUrl.replace('https://', '').split('.')[0]}/sql`);
    process.exit(1);
  }
}

main();

File: apply_migration_12.py
#!/usr/bin/env python3
"""
Apply migration 12: Fix material capacity MV unique index
This fixes the 500 error when updating expected materials.
"""

import os
import sys

try:
    import psycopg2
except ImportError:
    print("Error: psycopg2 not installed")
    print("Install it with: pip install psycopg2-binary")
    sys.exit(1)

# Load environment variables
from dotenv import load_dotenv
load_dotenv()

# Get Supabase project reference
supabase_url = os.getenv('SUPABASE_URL', '')
if not supabase_url:
    print("Error: SUPABASE_URL not set in .env")
    sys.exit(1)

project_ref = supabase_url.replace('https://', '').split('.')[0]

# Database connection string for Supabase
# Format: postgresql://postgres:[PASSWORD]@db.[PROJECT_REF].supabase.co:5432/postgres
db_password = os.getenv('SUPABASE_DB_PASSWORD', '')
if not db_password:
    db_password = input("Enter Supabase database password: ")

db_url = f"postgresql://postgres.{project_ref}:{db_password}@aws-0-ap-northeast-2.pooler.supabase.com:6543/postgres"

# Read migration SQL
migration_file = 'supabase/migration/12_fix_material_capacity_unique_index.sql'
with open(migration_file, 'r') as f:
    sql = f.read()

print(f"Connecting to Supabase project: {project_ref}")
print(f"Executing migration: {migration_file}")

try:
    # Connect to database
    conn = psycopg2.connect(db_url)
    conn.autocommit = True
    cur = conn.cursor()

    # Execute migration
    cur.execute(sql)

    print("\n✓ Migration executed successfully!")
    print("The 500 error should now be fixed.")

    # Close connection
    cur.close()
    conn.close()

except psycopg2.Error as e:
    print(f"\n✗ Database error: {e}")
    sys.exit(1)
except Exception as e:
    print(f"\n✗ Error: {e}")
    sys.exit(1)
